napari/utils/progress.py

```diff
@@ -0,0 +1,315 @@
+
+    return pbar
+
+
+def get_calling_function_name(max_depth: int):
+    """Inspect stack up to max_depth and return first function name outside of progress.py"""
+    for finfo in inspect.stack()[2:max_depth]:
+        if not finfo.filename.endswith("progress.py"):
+            return finfo.function
+
+    return None
+
+
+_tqdm_kwargs = {
+    p.name
+    for p in inspect.signature(tqdm.__init__).parameters.values()
+    if p.kind is not inspect.Parameter.VAR_KEYWORD and p.name != "self"
+}
+
+
+class progress(tqdm):
+    """This class inherits from tqdm and provides an interface for
+    progress bars in the napari viewer. Progress bars can be created
+    directly by wrapping an iterable or by providing a total number
+    of expected updates.
+
+    See tqdm.tqdm API for valid args and kwargs: https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
+    ...         sleep(delay)
+
+    it can also be used as a context manager:
+
+    >>> def long_running(steps=10, repeats=4, delay=0.1):
+    ...     with progress(range(steps)) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    or for manual updates:
+
+    >>> def manual_updates(total):
+    ...     pbr = progress(total=total)
+    ...     sleep(10)
+    ...     pbr.set_description("Step 1 Complete")
+    ...     pbr.update(1)
+
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+
+        # check if there's a napari viewer instance
+        viewer = get_viewer_instance()
+        if viewer is not None:
+            kwargs['gui'] = True
+
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        super().__init__(iterable, desc, total, *args, **kwargs)
+        if viewer is None:
+            return
+        self.viewer = viewer
+
+        self._pbar = get_pbar(viewer, **pbar_kwargs)
+        if self.total is not None:
+            self._pbar.setRange(self.n, self.total)
+            self._pbar._set_value(self.n)
+        else:
+            self._pbar.setRange(0, 0)
+
+        if desc:
+            self.set_description(desc)
+        else:
+            desc = get_calling_function_name(max_depth=5)
+            if desc:
+                self.set_description(desc)
+            else:
+                # TODO: pick a better default
+                self.set_description("Progress Bar")
+
+        self.show()
+        QApplication.processEvents()
+
+    def display(self, msg: str = None, pos: int = None) -> None:
+        """Update the display."""
+        if not self.viewer:
```

Comment:
I think this has the potential for an `AttributeError`, since `self.viewer` is defined _after_  `if viewer is None: return` in the `__init__`... so if we're hoping to be able to fall back to `tqdm` when there is no viewer, then we also need `self.viewer = None` somewhere.

Reply:
(yep... try `list(progress(range(10)))` without a viewer instance)